### PR TITLE
Remove redundant DeepCopy calls

### PR DESCRIPTION
**Pull request by xpivarc**
### What this PR does
Removes redundant `DeepCopy` calls because we now have https://github.com/kubevirt/kubevirt/pull/11265 

Fixes #

### Release note

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-controller/watch/vm.go [hunk 1]
```diff
@@ -712,8 +711,7 @@ func (c *VMController) VMIAffinityPatch(vm *virtv1.VirtualMachine, vmi *virtv1.V
 	var ops []string
 
 	if vm.Spec.Template.Spec.Affinity != nil {
-		vmAffinity := vm.Spec.Template.Spec.Affinity.DeepCopy()
-		vmAffinityJson, err := json.Marshal(vmAffinity)
+		vmAffinityJson, err := json.Marshal(vm.Spec.Template.Spec)
```
Comment (EdDev) [inline pkg/virt-controller/watch/vm.go:714]: You seem to be marshaling the whole `Spec` and not the affinity.

I guess this area is not unit tested.

Comment (vladikr) [reply pkg/virt-controller/watch/vm.go:714]: @xpivarc Functional test fails because of this change here.

Comment (xpivarc) [reply pkg/virt-controller/watch/vm.go]: Thanks, I was too eager with the delete button.

**Review by EdDev — COMMENTED** [review]
The new unit test fails due to that mock framework.

You could do this to avoid using it and still cover the logic:
Separate the calculation of the patch data from the actual patching, then test only the calculation logic.

**Comment by xpivarc** [comment]
> The new unit test fails due to that mock framework.
> 
> You could do this to avoid using it and still cover the logic: Separate the calculation of the patch data from the actual patching, then test only the calculation logic.

The test was flaky, it should be fixed now

File: pkg/virt-controller/watch/vm_test.go [hunk 2]
```diff
@@ -5335,6 +5335,64 @@ var _ = Describe("VirtualMachine", func() {
...
+							type P struct {
+								Op    string
+								Path  string
+								Value k8sv1.Affinity
+							}
+
+							patches := []P{}
+
+							Expect(json.Unmarshal(data, &patches)).To(Not(HaveOccurred()))
+							Expect(patches).To(HaveLen(1))
```
Comment (fossedihelm) [inline pkg/virt-controller/watch/vm_test.go:5382]: Why don't we just use patchOperation?
```suggestion
							patches := []patch.PatchOperation{}

							Expect(json.Unmarshal(data, &patches)).To(Not(HaveOccurred()))
							Expect(patches).To(HaveLen(1))
							aff := patches[0].Value.(k8sv1.Affinity)
```

**Review by fossedihelm — COMMENTED** [review]
Thank you @xpivarc!
Just one comment below!
Can I also ask you to add a proper PR description?

File: pkg/virt-controller/watch/vm_test.go (hunk 2 above)
Comment (xpivarc) [reply pkg/virt-controller/watch/vm_test.go:5382]: This doesn't work well. The `Unmarshal` will produce a `map[string]interface{}` which would then need to be converted to `Affinity` and then we could use the `equality.Semantic.DeepEqual`.

File: pkg/virt-controller/watch/vm_test.go [hunk 3]
```diff
@@ -5335,6 +5335,64 @@ var _ = Describe("VirtualMachine", func() {
...
+										MatchLabels: map[string]string{
+											"test": "nmc",
+										},
+									},
+								},
+							},
+						},
+					}
+					vm.Spec.Template.Spec.Affinity = affinity.DeepCopy()
+					Expect(vmiInformer.GetIndexer().Add(vmi)).To(Not(HaveOccurred()))
```
Comment (iholder101) [inline pkg/virt-controller/watch/vm_test.go:5369]:
```suggestion
					Expect(vmiInformer.GetIndexer().Add(vmi)).To(Succeed())
```

p.s. (nit) personally I don't like these one-liner execute+validate. I think the code is much clearer when they're separate to different lines. It's also easier to debug that way. But there are other opinions on that :)

File: pkg/virt-controller/watch/vm_test.go [hunk 4]
```diff
@@ -5335,6 +5335,64 @@ var _ = Describe("VirtualMachine", func() {
...
+						func(_, _, _ any, data []byte, _ any, _ ...any) (*virtv1.VirtualMachineInstance, error) {
+							type P struct {
+								Op    string
+								Path  string
+								Value k8sv1.Affinity
+							}
+
+							patches := []P{}
+
+							Expect(json.Unmarshal(data, &patches)).To(Not(HaveOccurred()))
```
Comment (iholder101) [inline pkg/virt-controller/watch/vm_test.go:5381]:
```suggestion
							Expect(json.Unmarshal(data, &patches)).To(Succeed())
```

File: pkg/virt-controller/watch/vm_test.go (hunk 2 above)
Comment (iholder101) [reply pkg/virt-controller/watch/vm_test.go:5382]: Can you please add a small code comment that explains that? I have also found it non-intuitive

Comment (iholder101) [inline pkg/virt-controller/watch/vm_test.go:5382]: Can you explain why this assumption has to be correct? IOW, why we should test it?

File: pkg/virt-controller/watch/vm_test.go [hunk 5]
```diff
@@ -5335,6 +5335,64 @@ var _ = Describe("VirtualMachine", func() {
 					Expect(vmConditionController.HasCondition(vm, virtv1.VirtualMachineRestartRequired)).To(BeTrue())
 				})
 			})
+
+			Context("Affinity", func() {
+				It("should work", func() {
```
Comment (iholder101) [inline pkg/virt-controller/watch/vm_test.go:5340]:
```suggestion
				It("should be live-updated", func() {
```

**Review by iholder101 — COMMENTED** [review]
Thanks @xpivarc! Looks great to me.

I left mostly tiny nits, feel free to ignore them.
I'll hold the PR just to give you a chance to have another look. Feel free to unhold when you feel the PR is ready.

/approve
/hold

**Comment by iholder101** [comment]
Also @xpivarc please update the PR description

File: pkg/virt-controller/watch/vm_test.go (hunk 3 above)
Comment (jean-edouard) [reply pkg/virt-controller/watch/vm_test.go:5369]: That suggestion is important for readability. We don't want the VMI add to "not have occurred", we want it to succeed.
However I tend to disagree with the one-liner comment. "expect add vmi to succeed" is easier to read than "err = add vmi ; expect err to not have occurred" IMO.

Comment (EdDev) [reply pkg/virt-controller/watch/vm_test.go:5369]: The one-liner `Succeed()` version is the preferred one indeed.

File: pkg/virt-controller/watch/vm_test.go (hunk 2 above)
Comment (xpivarc) [reply pkg/virt-controller/watch/vm_test.go:5382]: Added `By` to provide more context

**Comment by xpivarc** [comment]
@jean-edouard @EdDev @fossedihelm @iholder101 PTAL

**Comment by vladikr** [comment]
Thanks
/lgtm

File: pkg/virt-controller/watch/vm_test.go (hunk 3 above)
Comment (iholder101) [reply pkg/virt-controller/watch/vm_test.go]: Yeah I understand that I'm in the minority with this opinion :)

**Review by EdDev — APPROVED** [review]

**Comment by iholder101** [comment]
Thanks @xpivarc!
/hold cancel

**Pull request merged**